File: apps/ios/Wordocious/Sources/StoreManager.swift
import Foundation
import StoreKit

/// StoreKit 2 wrapper for the Pro subscription. Mirrors the web's Pro plans
/// (lib/payment/types.ts) and fulfillment (lib/payment/purchase-service.ts):
///   pro_monthly  $6.99  auto-renewable  → +30d, +4 shields
///   pro_yearly   $59.99 auto-renewable  → +365d, +4 shields
///   pro_day      $1.00  consumable      → +24h (stacks on existing window), no shields
///
/// Entitlement is written to the same `profiles` columns the web uses
/// (is_pro / pro_expires_at / streak_shields) so Pro is identical across clients.
@MainActor
final class StoreManager: ObservableObject {
    static let shared = StoreManager()

    enum Plan: String, CaseIterable {
        case monthly = "com.wordocious.app.pro_monthly"
        case yearly  = "com.wordocious.app.pro_yearly"
        case day     = "com.wordocious.app.pro_day"

        /// Days added per purchase (day pass handled with stacking semantics).
        var addedDays: Int { self == .yearly ? 365 : (self == .monthly ? 30 : 1) }
        var grantsShields: Bool { self != .day }
    }

    @Published private(set) var products: [Product] = []
    @Published private(set) var purchasingId: String?
    @Published var lastError: String?

    private var updatesTask: Task<Void, Never>?

    /// Transaction IDs whose one-time effects (shield grant, Day Pass +24h) have
    /// already been applied — so the .success and Transaction.updates paths
    /// can't double-apply the same delivery, and a re-delivered consumable
    /// after a crash can't stack a second 24h. A new billing period carries a
    /// new transaction.id, so this still grants shields once PER renewal.
    /// Persisted (bounded) so it survives relaunch.
    private static let processedKey = "storekit-processed-tx-ids"
    private var processedTxIDs: Set<UInt64> = {
        let raw = UserDefaults.standard.array(forKey: processedKey) as? [NSNumber] ?? []
        return Set(raw.map(\.uint64Value))
    }()

    private func hasProcessed(_ id: UInt64) -> Bool { processedTxIDs.contains(id) }
    private func markProcessed(_ id: UInt64) {
        processedTxIDs.insert(id)
        // Keep the most recent 500 — unbounded growth is the only downside.
        if processedTxIDs.count > 500 { processedTxIDs = Set(processedTxIDs.sorted().suffix(500)) }
        UserDefaults.standard.set(processedTxIDs.map { NSNumber(value: $0) }, forKey: Self.processedKey)
    }

    func product(for plan: Plan) -> Product? { products.first { $0.id == plan.rawValue } }

    private init() {}

    /// Call once at launch (after auth bootstrap). Loads products, reconciles any
    /// active subscription into the profile, and listens for renewals/revocations.
    func start() {
        updatesTask = Task { [weak self] in
            for await update in Transaction.updates {
                await self?.handle(verification: update, isNewPurchase: false)
            }
        }
        Task { await loadProducts() }
        Task { await syncCurrentEntitlements() }
    }

    func loadProducts() async {
        do {
            let loaded = try await Product.products(for: Plan.allCases.map(\.rawValue))
            // Stable display order: monthly, yearly, day.
            let order = Plan.allCases.map(\.rawValue)
            products = loaded.sorted { order.firstIndex(of: $0.id)! < order.firstIndex(of: $1.id)! }
        } catch {
            lastError = "Couldn't load products: \(error.localizedDescription)"
        }
    }

    // MARK: - Purchase

    func purchase(_ plan: Plan) async {
        guard let product = product(for: plan) else {
            lastError = "That plan isn't available right now."
            return
        }
        purchasingId = plan.rawValue
        defer { purchasingId = nil }
        do {
            // Tag the purchase with the Supabase user UUID so the server-side
            // App Store Server Notifications webhook can map the transaction to
            // this user (appAccountToken surfaces in the signed transaction).
            var options: Set<Product.PurchaseOption> = []
            if let session = try? await AuthService.shared.client.auth.session {
                options.insert(.appAccountToken(session.user.id))
            }
            let result = try await product.purchase(options: options)
            switch result {
            case .success(let verification):
                await handle(verification: verification, isNewPurchase: true)
            case .userCancelled:
                break
            case .pending:
                lastError = "Your purchase is pending approval."
            @unknown default:
                break
            }
        } catch {
            lastError = error.localizedDescription
        }
    }

    /// Re-sync from the App Store (Apple requires a Restore Purchases action).
    func restore() async {
        do {
            try await AppStore.sync()
            await syncCurrentEntitlements()
        } catch {
            lastError = error.localizedDescription
        }
    }

    // MARK: - Transaction handling

    private func handle(verification: VerificationResult<Transaction>, isNewPurchase: Bool) async {
        // Unverified is never trusted; surface it so a paying user isn't left
        // staring at a cleared spinner with no explanation.
        guard case .verified(let transaction) = verification else {
            if isNewPurchase { lastError = "Your purchase couldn't be verified. Try Restore Purchases." }
            return
        }
        guard let plan = Plan(rawValue: transaction.productID) else { await transaction.finish(); return }

        // Refund / revocation arrives on Transaction.updates. Revoke Pro rather
        // than fall through — a refunded sub whose expiry is still in the future
        // was otherwise re-affirmed as Pro. finish() only after a durable write.
        if transaction.revocationDate != nil || transaction.isUpgraded {
            if await AuthService.shared.revokePro() { markProcessed(transaction.id); await transaction.finish() }
            return
        }

        // The transaction must belong to the signed-in Supabase user. Without
        // this, account B on a device whose Apple ID owns A's subscription
        // would have Pro written onto B. Leave unfinished so it can be applied
        // once the owning account signs in.
        if !appAccountMatchesCurrentUser(transaction) { return }

        // A Day Pass already applied (crash re-delivery, or the .success/.updates
        // race) must not stack another 24h. Subscriptions re-reconcile harmlessly.
        if plan == .day && hasProcessed(transaction.id) { await transaction.finish(); return }

        let expiry = expiryDate(for: plan, transaction: transaction)
        // Shields once per delivery (per billing period): grant unless this exact
        // transaction was already processed — covers first purchase, each renewal,
        // and Ask-to-Buy approvals, without double-granting on the race.
        let shields = (plan.grantsShields && !hasProcessed(transaction.id)) ? 4 : 0
        let ok = await AuthService.shared.applyProGrant(expiresAt: expiry, addShields: shields)

        // Only mark + finish once the entitlement is durably written. On failure
        // the transaction stays unfinished → StoreKit re-delivers it via
        // Transaction.updates for retry (fixes the paid-but-unfulfilled Day Pass).
        if ok {
            markProcessed(transaction.id)
            await transaction.finish()
        }
    }

    /// On launch / restore: reflect any active auto-renewable entitlement into the
    /// profile without granting shields (reconciliation, not a new purchase).
    private func syncCurrentEntitlements() async {
        for await result in Transaction.currentEntitlements {
            guard case .verified(let transaction) = result,
                  let plan = Plan(rawValue: transaction.productID),
                  transaction.revocationDate == nil,
                  appAccountMatchesCurrentUser(transaction) else { continue }
            // Skip expired (StoreKit usually omits these, but be safe).
            if let exp = transaction.expirationDate, exp < Date() { continue }
            await AuthService.shared.syncProExpiry(expiresAt: expiryDate(for: plan, transaction: transaction))
        }
    }

    /// True when the transaction has no appAccountToken (legacy/unmapped — accept,
    /// matching the webhook's no-token no-op) or it equals the current user's id.
    private func appAccountMatchesCurrentUser(_ transaction: Transaction) -> Bool {
        guard let token = transaction.appAccountToken else { return true }
        guard let uid = AuthService.shared.profile?.id, let current = UUID(uuidString: uid) else { return false }
        return token == current
    }

    private func expiryDate(for plan: Plan, transaction: Transaction) -> Date {
        switch plan {
        case .monthly, .yearly:
            // Trust StoreKit's expiration for auto-renewables (keeps renewals in sync).
            return transaction.expirationDate ?? Date().addingTimeInterval(Double(plan.addedDays) * 86_400)
        case .day:
            // Day pass stacks on the existing Pro window (max semantics), like the web.
            let now = Date()
            let existing = AuthService.shared.profile?.proExpiresAt.flatMap(parseTimestamp)
            let base = max(existing ?? now, now)
            return base.addingTimeInterval(86_400)
        }
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/StoreManager.kt
package com.wordocious.app.data

import android.app.Activity
import android.content.Context
import com.android.billingclient.api.AcknowledgePurchaseParams
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.ConsumeParams
import com.android.billingclient.api.PendingPurchasesParams
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.PurchasesUpdatedListener
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.android.billingclient.api.acknowledgePurchase
import com.android.billingclient.api.consumePurchase
import com.android.billingclient.api.queryProductDetails
import com.android.billingclient.api.queryPurchasesAsync
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant

/**
 * Google Play Billing wrapper — Android analogue of the iOS StoreKit 2
 * StoreManager (apps/ios/Wordocious/Sources/StoreManager.swift) and the web's
 * lib/payment/purchase-service.ts fulfillment:
 *   pro_monthly  $6.99  auto-renewing sub  → +30d
 *   pro_yearly   $59.99 auto-renewing sub  → +365d
 *   pro_day      $1.00  consumable         → +24h stacked on any existing
 *                                            future expiry
 *
 * Entitlement is written to the same `profiles` columns every client uses
 * (is_pro / pro_expires_at) via AuthService.applyProGrant. The +4 streak
 * shields per paid period are NOT granted here — the Play RTDN webhook owns
 * them (see PAYMENTS_RUNBOOK "streak shields").
 * Everything is wrapped in runCatching — on emulators without Play services
 * or before the Play Console products exist, the screen simply keeps its
 * hardcoded fallback prices and purchase attempts surface a friendly error.
 */
object StoreManager {
    const val PRO_MONTHLY = "pro_monthly"
    const val PRO_YEARLY = "pro_yearly"
    const val PRO_DAY = "pro_day"

    /** Fallback display prices when Play products haven't loaded. */
    private val FALLBACK_PRICES = mapOf(
        PRO_MONTHLY to "$6.99",
        PRO_YEARLY to "$59.99",
        PRO_DAY to "$1",
    )

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private var billingClient: BillingClient? = null
    private var productDetails: Map<String, ProductDetails> = emptyMap()

    /**
     * Play orderIds whose one-time effect (Day Pass +24h) has been applied —
     * Android analogue of iOS StoreManager.processedTxIDs. Without it a Day Pass
     * re-delivered after a failed consume would stack a second 24h for free.
     *
     * It no longer tracks subscription orders: its other job was deciding when
     * to credit renewal shields client-side, and those are granted server-side
     * from the RTDN webhook now — which sees a renewal whether or not the app is
     * ever reopened.
     */
    private const val PREFS_NAME = "wordocious_store"
    private const val KEY_SEEN_ORDERS = "seen-order-ids"
    private var prefs: android.content.SharedPreferences? = null
    private var seenOrderIds: MutableSet<String> = mutableSetOf()

    private fun markOrderSeen(orderId: String) {
        seenOrderIds.add(orderId)
        // putStringSet must get a fresh copy — SharedPreferences may not persist
        // in-place mutations of the set it handed out.
        prefs?.edit()?.putStringSet(KEY_SEEN_ORDERS, HashSet(seenOrderIds))?.apply()
    }

    private val _prices = MutableStateFlow(FALLBACK_PRICES)
    /** productId → localized formattedPrice (falls back to the US prices above). */
    val prices: StateFlow<Map<String, String>> = _prices.asStateFlow()

    private val _purchasingId = MutableStateFlow<String?>(null)
    val purchasingId: StateFlow<String?> = _purchasingId.asStateFlow()

    private val _lastError = MutableStateFlow<String?>(null)
    val lastError: StateFlow<String?> = _lastError.asStateFlow()

    private val purchasesListener = PurchasesUpdatedListener { result, purchases ->
        _purchasingId.value = null
        when (result.responseCode) {
            BillingClient.BillingResponseCode.OK ->
                purchases?.forEach { p ->
                    scope.launch { runCatching { handlePurchase(p, isNewPurchase = true) } }
                }
            BillingClient.BillingResponseCode.USER_CANCELED -> Unit // not an error
            else -> _lastError.value =
                result.debugMessage.ifBlank { "Purchase failed (code ${result.responseCode})" }
        }
    }

    /** Call once at launch (next to AuthService.initialize()). Never throws. */
    fun start(context: Context) {
        runCatching {
            prefs = context.applicationContext
                .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            seenOrderIds = prefs?.getStringSet(KEY_SEEN_ORDERS, null)?.toMutableSet()
                ?: mutableSetOf()
            val client = BillingClient.newBuilder(context.applicationContext)
                .setListener(purchasesListener)
                .enablePendingPurchases(
                    PendingPurchasesParams.newBuilder().enableOneTimeProducts().build()
                )
                .build()
            billingClient = client
            client.startConnection(object : BillingClientStateListener {
                override fun onBillingSetupFinished(result: BillingResult) {
                    if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                        scope.launch { runCatching { queryProducts() } }
                        // Reconcile active subs at launch (iOS
                        // syncCurrentEntitlements parity). PurchasesUpdatedListener
                        // only fires on purchase ACTIONS, never on background
                        // renewals, so without this an Android sub's Pro is set
                        // once and never re-reflected until the user taps Restore.
                        scope.launch { runCatching { reconcileActiveSubs() } }
                        // A Day Pass left unconsumed because its grant didn't
                        // land is only ever redelivered by an INAPP query — the
                        // listener never re-fires for it. This is the retry path
                        // iOS gets for free from Transaction.updates.
                        scope.launch { runCatching { reconcileDayPasses() } }
                    }
                }

                override fun onBillingServiceDisconnected() {
                    // Play will reconnect lazily on the next purchase attempt.
                }
            })
        }
    }

    private suspend fun queryProducts() {
        val client = billingClient ?: return
        val subsParams = QueryProductDetailsParams.newBuilder()
            .setProductList(
                listOf(PRO_MONTHLY, PRO_YEARLY).map {
                    QueryProductDetailsParams.Product.newBuilder()
                        .setProductId(it)
                        .setProductType(BillingClient.ProductType.SUBS)
                        .build()
                }
            )
            .build()
        val inappParams = QueryProductDetailsParams.newBuilder()
            .setProductList(
                listOf(
                    QueryProductDetailsParams.Product.newBuilder()
                        .setProductId(PRO_DAY)
                        .setProductType(BillingClient.ProductType.INAPP)
                        .build()
                )
            )
            .build()

        // SUBS and INAPP can't share one query (Billing library constraint),
        // but they don't need to run back-to-back either.
        val loaded: List<ProductDetails> = coroutineScope {
            val subs = async {
                runCatching { client.queryProductDetails(subsParams).productDetailsList.orEmpty() }.getOrDefault(emptyList())
            }
            val inapp = async {
                runCatching { client.queryProductDetails(inappParams).productDetailsList.orEmpty() }.getOrDefault(emptyList())
            }
            subs.await() + inapp.await()
        }
        if (loaded.isEmpty()) return // keep fallbacks — products not created yet / no Play

        productDetails = loaded.associateBy { it.productId }
        _prices.value = FALLBACK_PRICES.mapValues { (id, fallback) ->
            loaded.firstOrNull { it.productId == id }?.displayPrice() ?: fallback
        }
    }

    private fun ProductDetails.displayPrice(): String? = when (productType) {
        BillingClient.ProductType.SUBS ->
            // Last pricing phase of the base offer = the recurring full price
            // (skips intro/free-trial phases if any are ever added).
            subscriptionOfferDetails?.firstOrNull()
                ?.pricingPhases?.pricingPhaseList?.lastOrNull()?.formattedPrice
        else -> oneTimePurchaseOfferDetails?.formattedPrice
    }

    // MARK: Purchase

    /** Launch the Play purchase sheet for pro_monthly / pro_yearly / pro_day. */
    fun purchase(activity: Activity, productId: String) {
        _lastError.value = null
        runCatching {
            val client = billingClient
            val details = productDetails[productId]
            if (client == null || !client.isReady || details == null) {
                _lastError.value = "That plan isn't available right now."
                return
            }
            val detailsParams = BillingFlowParams.ProductDetailsParams.newBuilder()
                .setProductDetails(details)
            if (details.productType == BillingClient.ProductType.SUBS) {
                val offerToken = details.subscriptionOfferDetails?.firstOrNull()?.offerToken
                if (offerToken == null) {
                    _lastError.value = "That plan isn't available right now."
                    return
                }
                detailsParams.setOfferToken(offerToken)
            }
            val flowParams = BillingFlowParams.newBuilder()
                .setProductDetailsParamsList(listOf(detailsParams.build()))
            // Tag the purchase with the Supabase user UUID — Play analogue of the
            // iOS appAccountToken (surfaces in RTDN / Play purchase records).
            AuthService.userId?.let { flowParams.setObfuscatedAccountId(it) }

            _purchasingId.value = productId
            val result = client.launchBillingFlow(activity, flowParams.build())
            if (result.responseCode != BillingClient.BillingResponseCode.OK) {
                _purchasingId.value = null
                _lastError.value =
                    result.debugMessage.ifBlank { "Couldn't start purchase (code ${result.responseCode})" }
            }
        }.onFailure {
            _purchasingId.value = null
            _lastError.value = it.message?.take(120) ?: "Purchase failed"
        }
    }

    /** Re-sync active subscriptions from Play (Restore Purchases). */
    fun restore() {
        _lastError.value = null
        scope.launch {
            runCatching {
                val client = billingClient
                if (client == null || !client.isReady) {
                    _lastError.value = "Google Play isn't available right now."
                    return@launch
                }
                val active = activeSubs(client)
                if (active.isEmpty()) {
                    _lastError.value = "No active subscription found."
                } else {
                    active.forEach { handlePurchase(it, isNewPurchase = false) }
                }
            }.onFailure { _lastError.value = it.message?.take(120) ?: "Restore failed" }
        }
    }

    /** Silent launch reconcile — same as restore() but no user-facing errors. */
    private suspend fun reconcileActiveSubs() {
        val client = billingClient ?: return
        if (!client.isReady) return
        activeSubs(client).forEach { runCatching { handlePurchase(it, isNewPurchase = false) } }
    }

    /** Silent launch retry for a Day Pass that was paid for but never granted. */
    private suspend fun reconcileDayPasses() {
        val client = billingClient ?: return
        if (!client.isReady) return
        client.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder().setProductType(BillingClient.ProductType.INAPP).build()
        ).purchasesList.forEach { runCatching { handlePurchase(it, isNewPurchase = false) } }
    }

    private suspend fun activeSubs(client: BillingClient): List<Purchase> =
        client.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder().setProductType(BillingClient.ProductType.SUBS).build()
        ).purchasesList.filter { it.purchaseState == Purchase.PurchaseState.PURCHASED }

    // MARK: Fulfillment

    private suspend fun handlePurchase(purchase: Purchase, isNewPurchase: Boolean) {
        if (purchase.purchaseState != Purchase.PurchaseState.PURCHASED) {
            // Play's PENDING state (slow card / cash payment / parental approval)
            // arrives on the OK branch — tell the user, same as iOS StoreKit's
            // `case .pending`. Only for a fresh purchase; a pending order found
            // by the silent launch reconcile must stay silent.
            if (isNewPurchase && purchase.purchaseState == Purchase.PurchaseState.PENDING) {
                _lastError.value = "Your purchase is pending approval."
            }
            return
        }
        val client = billingClient ?: return
        for (productId in purchase.products) {
            when (productId) {
                PRO_DAY -> {
                    // Re-delivery of an already-stacked Day Pass (a consume that
                    // failed and retried) must not add a second 24h — iOS
                    // hasProcessed parity.
                    val dayOrderId = purchase.orderId
                    var granted = dayOrderId != null && dayOrderId in seenOrderIds
                    if (!granted) {
                        // Stack 24h on any existing future Pro window (max
                        // semantics — same as iOS StoreManager.expiryDate(.day)
                        // and the web). The grant needs the signed-in profile as
                        // its base; without one, leave the purchase untouched so
                        // a later reconcile retries it.
                        val now = Instant.now()
                        val loaded = AuthService.profile.value
                        if (loaded != null) {
                            // AuthService.parseTimestamp handles PostgREST's offset
                            // format — Instant.parse rejected it, so stacking never
                            // saw the existing window and every Day Pass reset to
                            // now+24h.
                            val existing = loaded.proExpiresAt?.let { AuthService.parseTimestamp(it) }
                            val base = if (existing != null && existing.isAfter(now)) existing else now
                            val target = base.plus(Duration.ofDays(1))
                            AuthService.applyProGrant(target.toString())
                            // iOS only finishes a Day Pass transaction once the
                            // entitlement is durably written, leaving it
                            // unfinished for retry otherwise (StoreManager.swift
                            // `case .failed: break`). applyProGrant re-reads the
                            // row from the server, so a stored expiry that
                            // reached the target IS that confirmation.
                            val written = AuthService.profile.value?.proExpiresAt
                                ?.let { AuthService.parseTimestamp(it) }
                            granted = written != null && !written.isBefore(target.minusSeconds(5))
                            if (granted) dayOrderId?.let { markOrderSeen(it) }
                        }
                    }
                    // Consumable: consume so it can be bought again — but only
                    // after a confirmed grant. Consuming first would burn the
                    // purchase token (Play never redelivers it) and the $1 would
                    // be lost outright if the write then failed.
                    if (granted) {
                        runCatching {
                            client.consumePurchase(
                                ConsumeParams.newBuilder()
                                    .setPurchaseToken(purchase.purchaseToken)
                                    .build()
                            )
                        }
                    }
                }
                PRO_MONTHLY, PRO_YEARLY -> {
                    if (!purchase.isAcknowledged) {
                        runCatching {
                            client.acknowledgePurchase(
                                AcknowledgePurchaseParams.newBuilder()
                                    .setPurchaseToken(purchase.purchaseToken)
                                    .build()
                            )
                        }
                    }
                    val days = if (productId == PRO_YEARLY) 365L else 30L
                    val now = Instant.now()
                    val existing = AuthService.profile.value?.proExpiresAt
                        ?.let { AuthService.parseTimestamp(it) }
                    val floor = now.plus(Duration.ofDays(days))

                    if (isNewPurchase) {
                        // A real payment: extend by the plan's length. This used
                        // to write now+days flat, which threw away every day the
                        // player already held from other sources — referral
                        // rewards, admin comps, stacked Day Passes. Someone with
                        // 97 banked days who bought monthly landed on 30 and lost
                        // the other 67, with no record to restore from.
                        // applyProGrant takes the later of this and the stored
                        // window, so the purchase can only ever add.
                        AuthService.applyProGrant(floor.toString())
                    } else if (existing == null || !existing.isAfter(now)) {
                        // RECONCILE/RESTORE with no live window. Play only returns
                        // a sub here while it's ACTIVE, but Purchase carries no
                        // expiry — purchaseTime is the ORIGINAL purchase, so
                        // purchaseTime+days is months in the PAST for a renewed
                        // sub, and writing that made Restore revoke Pro (F5).
                        // Re-establish with a conservative now+days; the RTDN
                        // webhook / server sweep is the real authority once live.
                        AuthService.applyProGrant(floor.toString())
                    }
                    // else: the stored window is already live, so leave it alone.
                    // Writing now+days on every launch would push a monthly
                    // subscriber's expiry out another 30 days each time the app
                    // opened — and now that the webhook never shrinks an expiry
                    // either, that inflation would be permanent free Pro.
                    //
                    // No shield grant on any of these paths. The +4 per billing
                    // period comes from the RTDN webhook (PURCHASED / RENEWED /
                    // RECOVERED), which is the only side that sees a renewal for
                    // a player who never reopens the app, and the only side a
                    // user can't fake by PATCHing their own profiles row. This
                    // client used to add its own +4 on top, so every subscriber
                    // was quietly getting 8 shields a period instead of 4.
                }
            }
        }
    }

    fun clearError() {
        _lastError.value = null
    }
}
